guava/src/com/google/common/util/concurrent/MoreExecutors.java

```diff
@@ -841,4 +910,53 @@ static ScheduledExecutorService renamingDecorator(final ScheduledExecutorService
       }
     };
   }
+
+  /**
+   * Shuts down the given executor gradually, first disabling new submissions and later cancelling
+   * existing tasks.
+   *
+   * <p>The method takes the following steps:
+   * <ol>
+   *  <li>calls {@link ExecutorService#shutdown()}, disabling acceptance of new submitted tasks.
+   *  <li>waits for half of the specified timeout.
+   *  <li>if the timeout expires, it calls {@link ExecutorService#shutdownNow()}, cancelling
+   *  pending tasks and interrupting running tasks.
+   *  <li>waits for the other half of the specified timeout.
+   * </ol>
+   *
+   * <p>If, at any step of the process, the given executor is terminated or the calling thread is
+   * interrupted, the method calls {@link ExecutorService#shutdownNow()}, cancelling
+   * pending tasks and interrupting running tasks.
+   *
+   * @param service the {@code ExecutorService} to shut down
+   * @param timeout the maximum time to wait for the {@code ExecutorService} to terminate
+   * @param unit the time unit of the timeout argument
+   * @return {@code true} if the pool was terminated successfully, {@code false} if the
+   *     {@code ExecutorService} could not terminate <b>or</b> the thread running this method
+   *     is interrupted while waiting for the {@code ExecutorService} to terminate
+   * @since 17.0
+   */
+  @Beta
+  public static boolean shutdownAndAwaitTermination(
+      ExecutorService service, long timeout, TimeUnit unit) {
+    checkNotNull(unit);
+    // Disable new tasks from being submitted
+    service.shutdown();
+    try {
+      long halfTimeoutNanos = TimeUnit.NANOSECONDS.convert(timeout, unit) / 2;
+      // Wait for half the duration of the timeout for existing tasks to terminate
+      if (!service.awaitTermination(halfTimeoutNanos, TimeUnit.NANOSECONDS)) {
+        // Cancel currently executing tasks
+        service.shutdownNow();
+        // Wait the other half of the timeout for tasks to respond to being cancelled
+        service.awaitTermination(halfTimeoutNanos, TimeUnit.NANOSECONDS);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Check the return value of `awaitTermination()` to determine if the operation timed out while waiting for other threads to stop executing. Handle when `awaitTermination()` times out. See the following diffs on GitHub that make this code change: [diff1](https://github.com/voldemort/voldemort/commit/e44ec88c57c1293a5b71598e0ad0fed403387744#diff-c220173cebaeedd4ef7bad06e679a0938710546ed77afac1d37a3ddf25e30e97L533), [diff2](https://github.com/utPLSQL/utPLSQL-cli/commit/079e021a747fe3829f868ad08c95a6eef2cc2f1c#diff-8a4bc1fbdd325f803130235d0cc70d6ae0d4b491b7440263c4f8a55452fd45eeL204)

Learn more: [ExecutorService.awaitTermination in the Java Platform, Standard Edition 8 API Specification](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ExecutorService.html#awaitTermination-long-java.util.concurrent.TimeUnit-)